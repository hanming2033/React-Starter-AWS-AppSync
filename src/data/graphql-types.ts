/* tslint:disable */
//  This file was automatically generated and should not be edited.

export interface GetLocalStatesQuery {
  forms:  {
    __typename: "forms",
    input_Email: string,
  } | null,
  nav:  {
    __typename: "nav",
    nextPath: string,
  } | null,
};
