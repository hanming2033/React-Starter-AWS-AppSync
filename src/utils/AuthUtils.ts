import { QueryResult } from 'react-apollo'
import { GetLocalStatesQuery } from '../data/graphql-types'

export const updateCacheForm = (qryRes: QueryResult<GetLocalStatesQuery>, propName: string, value: string) => {
  if (qryRes.data && qryRes.data.forms) {
    const newData: GetLocalStatesQuery = {
      ...qryRes.data,
      forms: {
        ...qryRes.data.forms,
        [propName]: value
      }
    }
    qryRes.client.writeData({ data: newData })
  }
}
