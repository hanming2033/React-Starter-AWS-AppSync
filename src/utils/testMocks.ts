import { GetLocalStatesQuery } from '../data/graphql-types'
import { QueryResult } from 'react-apollo'
import { ApolloError } from 'apollo-client'
import { client } from '../AWSApolloClient'

export const getLocalStateData: GetLocalStatesQuery = {
  auth: {
    __typename: 'auth',
    isAuthenticated: false
  },
  forms: {
    __typename: 'forms',
    input_Email: ''
  },
  nav: {
    __typename: 'nav',
    nextPath: ''
  }
}

export const qryRes: QueryResult<GetLocalStatesQuery> = {
  client,
  data: undefined,
  error: undefined,
  loading: false,
  networkStatus: 7,
  startPolling: jest.fn(),
  stopPolling: jest.fn(),
  subscribeToMore: jest.fn(),
  variables: {},
  refetch: jest.fn(),
  fetchMore: jest.fn(),
  updateQuery: jest.fn()
}

export const apolloError: ApolloError = {
  name: 'error',
  message: 'error',
  graphQLErrors: [{ name: 'error', message: 'error' }],
  networkError: null,
  extraInfo: ''
}
