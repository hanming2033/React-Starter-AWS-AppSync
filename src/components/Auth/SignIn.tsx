import * as React from 'react'
import { Form, Field, Formik, FormikActions } from 'formik'
import { Query, QueryResult } from 'react-apollo'
import { GetLocalStatesQuery } from '../../data/graphql-types'
import { GET_LOCAL_STATES } from '../../data/actions/Queries'
import * as yup from 'yup'
import { AuthProxy } from './AuthProxies/AuthProxy'
import { TtoComp, TsetAuth } from './AuthenticatorRouter'
import { verifyUser } from './AuthProxies/verifyUser'
import { updateCacheForm } from '../../utils/AuthUtils'

export interface ISignInProps {
  toComp: TtoComp
  setAuth: TsetAuth
}

export interface ISignInState {}

// *1 define interface for form values
interface ISigninFormValues {
  email: string
  password: string
}

type TsubmitFn = (
  values: ISigninFormValues,
  formikBag: FormikActions<ISigninFormValues>,
  qryRes: QueryResult<GetLocalStatesQuery>,
  props: ISignInProps
) => void

// *2 define yup schema for form validation
const schemaSignup = yup.object().shape({
  email: yup
    .string()
    .email('Not a valid email')
    .required('Email is required'),
  password: yup
    .string()
    .min(6, 'Minimum 6 characters')
    .required('Password is required')
})

// *3 create actual formik form component
export const FormikSignIn = (qryRes: QueryResult<GetLocalStatesQuery>, submit: TsubmitFn, props: ISignInProps) => (
  <Formik
    // get current user email from global state first
    initialValues={{
      email: (qryRes.data && qryRes.data.forms && (qryRes.data.forms.input_Email as string)) || '',
      password: ''
    }}
    validationSchema={schemaSignup}
    onSubmit={(values, formikBag) => submit(values, formikBag, qryRes, props)}
    render={formikProps => (
      <Form>
        <Field name="email" placeholder="Email" />
        {formikProps.touched.email && <span>{formikProps.errors.email}</span>}
        <br />
        <Field type="password" name="password" placeholder="Password" />
        {formikProps.touched.password && <span>{formikProps.errors.password}</span>}
        <br />
        <button type="submit" disabled={formikProps.isSubmitting}>
          Sign In
        </button>
      </Form>
    )}
  />
)

// *4 create onsubmit method
const loginSubmit = async (
  values: ISigninFormValues,
  formikBag: FormikActions<ISigninFormValues>,
  qryRes: QueryResult<GetLocalStatesQuery>,
  props: ISignInProps
) => {
  formikBag.setSubmitting(true)
  // store username in apollo link state on submit
  updateCacheForm(qryRes, 'input_Email', values.email)

  const res = await AuthProxy.signIn(values.email, values.password)
  if (res.data) {
    if (res.data.challengeName === 'SMS_MFA' || res.data.challengeName === 'SOFTWARE_TOKEN_MFA') {
      props.toComp('confirmSignIn') // Future: check if mfa works
    } else if (res.data.challengeName === 'NEW_PASSWORD_REQUIRED') {
      props.toComp('requireNewPassword', res.data)
    } else {
      const verificationDetail = await verifyUser(res.data)
      props.setAuth(verificationDetail)
    }
  } else if (res.error) {
    formikBag.setSubmitting(false)
    formikBag.setFieldValue('password', '', false)
    formikBag.setErrors({
      email: res.error.code ? res.error.message : '',
      password: res.error.code === 'NotAuthorizedException' ? res.error.message : ''
    })
    if (res.error.code === 'UserNotConfirmedException') {
      props.toComp('confirmSignUp')
    }
  }
}

// *Component
class Signin extends React.Component<ISignInProps, ISignInState> {
  public render() {
    return (
      <Query<GetLocalStatesQuery> query={GET_LOCAL_STATES} fetchPolicy="no-cache">
        {qryRes => {
          if (qryRes.loading) return <h1>Loading...</h1>
          if (qryRes.error || !qryRes.data) return <h1>Error...</h1>
          return (
            <>
              <h1>Sign In</h1>
              {FormikSignIn(qryRes, loginSubmit, this.props)}
              <button onClick={() => this.props.toComp('forgotPassword')}>Forgot Password</button>
              <button onClick={() => this.props.toComp('signUp')}>Go To SignUp</button>
            </>
          )
        }}
      </Query>
    )
  }
}

export default Signin
