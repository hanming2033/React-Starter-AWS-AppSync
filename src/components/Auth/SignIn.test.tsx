import * as React from 'react'
import * as enzyme from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'
import SignIn, { FormikSignIn } from './SignIn'
import { Query, QueryResult } from 'react-apollo'

import { qryRes, apolloError, getLocalStateData } from '../../utils/testMocks'
import { GetLocalStatesQuery } from '../../data/graphql-types'
import wait from 'waait'

enzyme.configure({ adapter: new Adapter() })

// * always try to test component rendering instead of state unless rendering is happening in child components
// * always try to test on interaction instead of state change unless interaction happens in child components
// * Group by feature. Each feature test on rendering and its interaction
// * -- rendering: how props and state(prefer interaction on component unless interaction happen in children) affect this component
// * -- interaction: how feature's interaction affect itself, other component, state(prefer rendering unless rendering in children), prop method call

describe('<SignIn /> Main Suite', () => {
  it('should render Loading... if local state has loading', () => {
    const toComp = jest.fn()
    const setAuth = jest.fn()
    const queryWrapper = enzyme.shallow(<SignIn toComp={toComp} setAuth={setAuth} />)
    const newRes: QueryResult<GetLocalStatesQuery> = { ...qryRes, loading: true }
    const wrapper = enzyme.shallow(queryWrapper.find(Query).prop('children')(newRes) as React.ReactElement<{}>)
    expect(wrapper.find('[children="Loading..."]').length).toBe(1)
  })
  it('should render Error... if local state has error or no data', () => {
    const toComp = jest.fn()
    const setAuth = jest.fn()
    const queryWrapper = enzyme.shallow(<SignIn toComp={toComp} setAuth={setAuth} />)
    const newRes: QueryResult<GetLocalStatesQuery> = { ...qryRes, error: apolloError }
    const wrapper = enzyme.shallow(queryWrapper.find(Query).prop('children')(newRes) as React.ReactElement<{}>)
    expect(wrapper.find('[children="Error..."]').length).toBe(1)
  })
  describe('h1 Sign In Suite', () => {
    it('should render h1 with text Sign In', () => {
      const toComp = jest.fn()
      const setAuth = jest.fn()
      const queryWrapper = enzyme.shallow(<SignIn toComp={toComp} setAuth={setAuth} />)
      const newRes: QueryResult<GetLocalStatesQuery> = { ...qryRes, data: getLocalStateData }
      const wrapper = enzyme.shallow(queryWrapper.find(Query).prop('children')(newRes) as React.ReactElement<{}>)
      expect(wrapper.find('[children="Sign In"]').length).toBe(1)
    })
  })
  describe('Forgot Password Button Suite', () => {
    it('should render Forgot Password', () => {
      const toComp = jest.fn()
      const setAuth = jest.fn()
      const queryWrapper = enzyme.shallow(<SignIn toComp={toComp} setAuth={setAuth} />)
      const newRes: QueryResult<GetLocalStatesQuery> = { ...qryRes, data: getLocalStateData }
      const wrapper = enzyme.shallow(queryWrapper.find(Query).prop('children')(newRes) as React.ReactElement<{}>)
      expect(wrapper.find('[children="Forgot Password"]').length).toBe(1)
    })
    it('should call toComp(propMethod) with args: forgotPassword', () => {
      const toComp = jest.fn()
      const setAuth = jest.fn()
      const queryWrapper = enzyme.shallow(<SignIn toComp={toComp} setAuth={setAuth} />)
      const newRes: QueryResult<GetLocalStatesQuery> = { ...qryRes, data: getLocalStateData }
      const wrapper = enzyme.shallow(queryWrapper.find(Query).prop('children')(newRes) as React.ReactElement<{}>)
      wrapper.find('[children="Forgot Password"]').simulate('click')
      expect(toComp.mock.calls[0][0]).toBe('forgotPassword')
    })
  })
  describe('To Sign Up Button Suite', () => {
    it('should render Go To SignUp', () => {
      const toComp = jest.fn()
      const setAuth = jest.fn()
      const queryWrapper = enzyme.shallow(<SignIn toComp={toComp} setAuth={setAuth} />)
      const newRes: QueryResult<GetLocalStatesQuery> = { ...qryRes, data: getLocalStateData }
      const wrapper = enzyme.shallow(queryWrapper.find(Query).prop('children')(newRes) as React.ReactElement<{}>)
      expect(wrapper.find('[children="Go To SignUp"]').length).toBe(1)
    })
    it('should call toComp(propMethod) with args: signUp', () => {
      const toComp = jest.fn()
      const setAuth = jest.fn()
      const queryWrapper = enzyme.shallow(<SignIn toComp={toComp} setAuth={setAuth} />)
      const newRes: QueryResult<GetLocalStatesQuery> = { ...qryRes, data: getLocalStateData }
      const wrapper = enzyme.shallow(queryWrapper.find(Query).prop('children')(newRes) as React.ReactElement<{}>)
      wrapper.find('[children="Go To SignUp"]').simulate('click')
      expect(toComp.mock.calls[0][0]).toBe('signUp')
    })
  })
})

// * always try to test component rendering instead of state unless rendering is happening in child components
// * always try to test on interaction instead of state change unless interaction happens in child components
// * Group by feature. Each feature test on rendering and its interaction
// * -- rendering: how props and state(prefer interaction on component unless interaction happen in children) affect this component
// * -- interaction: how feature's interaction affect itself, other component, state(prefer rendering unless rendering in children), prop method call

describe('<Formik SignIn /> Main Suite(Mounting)', () => {
  describe('Email Input Suite', async () => {
    const subtmitFn = jest.fn()
    it('should render "" if qyrRes has "" email field', () => {
      const newRes: QueryResult<GetLocalStatesQuery> = { ...qryRes, data: getLocalStateData }
      const mounted = enzyme.mount(FormikSignIn(newRes, subtmitFn, { toComp: jest.fn(), setAuth: jest.fn() }))
      const inputEmail = mounted.find('input[name="email"]')
      expect(inputEmail.props().value).toBe('')
    })
    it('should render "hanming2033@gmail.com" if qyrRes has "hanming2033@gmail.com" email field', () => {
      if (!getLocalStateData || !getLocalStateData.forms) return
      const newRes: QueryResult<GetLocalStatesQuery> = {
        ...qryRes,
        data: { ...getLocalStateData, forms: { ...getLocalStateData.forms, input_Email: 'hanming2033@gmail.com' } }
      }
      const mounted = enzyme.mount(FormikSignIn(newRes, subtmitFn, { toComp: jest.fn(), setAuth: jest.fn() }))
      const inputEmail = mounted.find('input[name="email"]')
      expect(inputEmail.props().value).toBe('hanming2033@gmail.com')
    })
    it('should render "abc" on input change "abc"', () => {
      const newRes: QueryResult<GetLocalStatesQuery> = { ...qryRes, data: getLocalStateData }
      const mounted = enzyme.mount(FormikSignIn(newRes, subtmitFn, { toComp: jest.fn(), setAuth: jest.fn() }))
      mounted.find('input[name="email"]').simulate('change', { persist: () => undefined, target: { name: 'email', value: 'abc' } })
      expect(mounted.find('input[name="email"]').props().value).toBe('abc')
    })
    it('should render "required error" on input change "" and blur', async () => {
      const newRes: QueryResult<GetLocalStatesQuery> = { ...qryRes, data: getLocalStateData }
      const mounted = enzyme.mount(FormikSignIn(newRes, subtmitFn, { toComp: jest.fn(), setAuth: jest.fn() }))
      mounted.find('input[name="email"]').simulate('change', { persist: () => undefined, target: { name: 'email', value: '' } })
      mounted.find('input[name="email"]').simulate('blur', { persist: () => undefined, target: { name: 'email' } })
      await wait(0)
      mounted.update()
      expect(mounted.find('[children="Email is required"]').length).toBe(1)
    })
    it('should render "not valid email error" on input change "abc" and blur', async () => {
      const newRes: QueryResult<GetLocalStatesQuery> = { ...qryRes, data: getLocalStateData }
      const mounted = enzyme.mount(FormikSignIn(newRes, subtmitFn, { toComp: jest.fn(), setAuth: jest.fn() }))
      mounted.find('input[name="email"]').simulate('change', { persist: () => undefined, target: { name: 'email', value: 'abc' } })
      mounted.find('input[name="email"]').simulate('blur', { persist: () => undefined, target: { name: 'email' } })
      await wait(0)
      mounted.update()
      expect(mounted.find('[children="Not a valid email"]').length).toBe(1)
    })
  })
  describe('Password Input Suite', () => {
    const subtmitFn = jest.fn()
    const newRes: QueryResult<GetLocalStatesQuery> = { ...qryRes, data: getLocalStateData }
    const mounted = enzyme.mount(FormikSignIn(newRes, subtmitFn, { toComp: jest.fn(), setAuth: jest.fn() }))
    it('should render "" if qyrRes has "" password field', () => {
      expect(mounted.find('input[name="password"]').props().value).toBe('')
    })
    it('should render "abc" on input change "abc"', () => {
      mounted.find('input[name="password"]').simulate('change', { persist: () => undefined, target: { name: 'password', value: 'abc' } })
      expect(mounted.find('input[name="password"]').props().value).toBe('abc')
    })
    it('should render "required error" on input change "" and blur', async () => {
      mounted.find('input[name="password"]').simulate('change', { persist: () => undefined, target: { name: 'password', value: '' } })
      mounted.find('input[name="password"]').simulate('blur', { persist: () => undefined, target: { name: 'password' } })
      await wait(0)
      mounted.update()
      expect(mounted.find('[children="Password is required"]').length).toBe(1)
    })
    it('should render "too short error" on input change "abc" and blur', async () => {
      mounted.find('input[name="password"]').simulate('change', { persist: () => undefined, target: { name: 'password', value: 'abc' } })
      mounted.find('input[name="password"]').simulate('blur', { persist: () => undefined, target: { name: 'password' } })
      await wait(0)
      mounted.update()
      expect(mounted.find('[children="Minimum 6 characters"]').length).toBe(1)
      console.log(mounted.debug())
    })
  })
  describe('Submit Button Suite', () => {
    it('should render a button with "Sign In"', () => {
      const subtmitFn = jest.fn()
      const newRes: QueryResult<GetLocalStatesQuery> = { ...qryRes, data: getLocalStateData }
      const mounted = enzyme.mount(FormikSignIn(newRes, subtmitFn, { toComp: jest.fn(), setAuth: jest.fn() }))
      expect(mounted.find('button').text()).toBe('Sign In')
    })
    it('should have disable state = false initially', () => {
      const subtmitFn = jest.fn()
      const newRes: QueryResult<GetLocalStatesQuery> = { ...qryRes, data: getLocalStateData }
      const mounted = enzyme.mount(FormikSignIn(newRes, subtmitFn, { toComp: jest.fn(), setAuth: jest.fn() }))
      expect(mounted.find('button').props().disabled).toBe(false)
    })
    it('should call submitFn on click', () => {
      const subtmitFn = jest.fn()
      const newRes: QueryResult<GetLocalStatesQuery> = { ...qryRes, data: getLocalStateData }
      const mounted = enzyme.mount(FormikSignIn(newRes, subtmitFn, { toComp: jest.fn(), setAuth: jest.fn() }))
      console.log(mounted.state())
      mounted.find('button').simulate('click')
      expect(subtmitFn).toBeCalled()
    })
    it('should disable the button on click', () => {})
  })
})

// describe('Mounting', () => {
//   it('should render form', async () => {
//     const subtmitFn = jest.fn()
//     const newRes: QueryResult<GetLocalStatesQuery> = { ...qryRes, data: getLocalStateData }
//     const mounted = enzyme.mount(FormikSignIn(newRes, subtmitFn, { toComp: jest.fn(), setAuth: jest.fn() }))
//     mounted
//       .find('input[name="email"]')
//       .simulate('change', { persist: () => undefined, target: { name: 'email', value: 'hanming@gmail.com' } })
//     console.log(mounted.state())
//     mounted.find('input[name="email"]').simulate('blur', { persist: () => undefined, target: { name: 'email' } })
//     await wait(0)
//     console.log(mounted.state())
//     // mounted.find('button').simulate('click')
//     mounted.update()
//     console.log(mounted.debug())
//   })
// })
