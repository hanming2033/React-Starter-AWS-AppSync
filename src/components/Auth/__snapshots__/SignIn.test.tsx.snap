// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Formik SignIn /> Main Suite(Mounting) Email Input Suite should render "" if qyrRes has "" email field 1`] = `
<input
  name="email"
  onBlur={[Function]}
  onChange={[Function]}
  placeholder="Email"
  value=""
/>
`;

exports[`<Formik SignIn /> Main Suite(Mounting) Email Input Suite should render "hanming2033@gmail.com" if qyrRes has "hanming2033@gmail.com" email field 1`] = `
<input
  name="email"
  onBlur={[Function]}
  onChange={[Function]}
  placeholder="Email"
  value="hanming2033@gmail.com"
/>
`;

exports[`<Formik SignIn /> Main Suite(Mounting) Password Input Suite should render password input snapshot on default qryRes prop 1`] = `
<input
  name="password"
  onBlur={[Function]}
  onChange={[Function]}
  placeholder="Password"
  type="password"
  value=""
/>
`;

exports[`<Formik SignIn /> Main Suite(Mounting) Submit Button Suite should render a button snapshot" 1`] = `
<button
  disabled={false}
  type="submit"
>
  Sign In
</button>
`;

exports[`<SignIn /> Main Suite Forgot Password Button Suite should render Forgot Password 1`] = `
<button
  onClick={[Function]}
>
  Forgot Password
</button>
`;

exports[`<SignIn /> Main Suite h1 Sign In Suite should render h1 with text Sign In 1`] = `
<h1>
  Sign In
</h1>
`;

exports[`<SignIn /> Main Suite should render Error... if local state has error or no data 1`] = `
<h1>
  Error...
</h1>
`;

exports[`<SignIn /> Main Suite should render Loading... if local state has loading 1`] = `
<h1>
  Loading...
</h1>
`;
